src/utils.c

```diff
@@ -181,6 +185,33 @@ bool is_heap(void *addr)
     return p > top_of_kernel_data && p < top_of_ram;
 }
 
+bool is_idevice(void)
+{
+    return chip_id == 0x8960 || chip_id == 0x7000 || chip_id == 0x7001 || chip_id == 0x8000 ||
+           chip_id == 0x8001 || chip_id == 0x8003 || chip_id == 0x8010 || chip_id == 0x8011 ||
+           chip_id == 0x8012 || chip_id == 0x8015; /* Not the most elegant but works well... */
+}
+
+bool supports_arch_retention(void)
+{
+    return mrs(AIDR_EL1) & AIDR_EL1_ARCH_RETENTION;
+}
+
+bool supports_fipi(void)
+{
+    return (!is_idevice() || chip_id == 0x8015);
```

Comment:
Probably just test the boot CPU part number for this? See the test for architectural ACTLR_EL12 that got added recently, we can have more flags like that.